[FEATURE] Write static headers to to the server config file

zero-24 opened the pull request:
Hi,

this is the code for the next feature update of the plugin adding support to write the static headers to the .htaccess and web.config files without interaction of the user as well as updating the values when already set when the user saves the extension from the backend.

### New Language strings

Here are the new language strings shipped with this feature please let me know whether any of this is unknown or can be improved. When you get the time please send the new translated strings to this branch here: https://github.com/zero-24/plg_system_httpheader/tree/saveToServerConfigFile or post them here as comment and i can commit them on your behalf.
Please let me know when you have any questions or remarks on the strings.

cc @pe7er (ne-NL) @YGomiero (fr-FR) @brianteeman (en-GB)

```ini
PLG_SYSTEM_HTTPHEADER_WRITE_STATIC_HEADERS="Write headers to the configuration file"
PLG_SYSTEM_HTTPHEADER_WRITE_STATIC_HEADERS_DESC="When enabled the headers (excluding the Content-Security-Policy) will be added to the existing server configuration file as soon as you save this extension. This way the headers will be enforced from the server side. At this time we only support .htaccess (Apache) and web.config (IIS) files."
PLG_SYSTEM_HTTPHEADER_MESSAGE_STATICHEADERS_NOT_WRITTEN_NO_SERVER_CONFIGFILE_FOUND="We couldn't find any active .htaccess or web.config file to apply the rules to. Please first rename the htaccess.txt to .htaccess or web.config.txt to web.config and make sure the configuration files works."
PLG_SYSTEM_HTTPHEADER_MESSAGE_STATICHEADERS_WRITTEN="All static headers have been written to the server configuration file."
PLG_SYSTEM_HTTPHEADER_MESSAGE_STATICHEADERS_NOT_WRITTEN="Your %1$s file is not writable or there was a problem creating the file. You will have to upload the following code by hand. Select the following code and then paste into a new text file. Name this file '%1$s' and upload it to your site root folder.<br><pre>%2$s</pre>"
```

### How to test

Install this package representing the latest development version and report any issues remarks here: 
[plg_system_httpheader.zip](https://github.com/zero-24/plg_system_httpheader/files/2876974/plg_system_httpheader.zip)

I would like to ask @snipersister and @yvesh to take a look from the PHP side and give this an review, but also any other kind of feedback is appreciated :)

### Additional note

When ready the plan is this to be merged up to the version of this plugin shipped with Joomla 4.0 too.

Thanks for your support 👍

zero-24 committed:
implement the first steps for the write to the server config ffiles

zero-24 committed:
web.config part and some cleanup

zero-24 committed:
implement htaccess generator

zero-24 committed:
complete htaccess code

zero-24 committed:
add en-GB language string and UI option

zero-24 committed:
apply xml changes

zero-24 committed:
fix validation rule and wrong variable names

zero-24 committed:
some coderstyle changes

zero-24 committed:
some more fixes

zero-24 committed:
improve language string about the error message

zero-24 committed:
alpha order language strings

zero-24 committed:
2018 -> 2019 && .sys.ini vs .ini

zero-24 committed:
de-DE translation

brianteeman commented:
```ini
PLG_SYSTEM_HTTPHEADER_WRITE_STATIC_HEADERS="Write headers to the configuration file"
PLG_SYSTEM_HTTPHEADER_WRITE_STATIC_HEADERS_DESC="When enabled the headers (excluding the Content-Security-Policy) will be added to the existing server configuration file as soon as you save this extension. This way the headers will be enforced from the server side. At this time we only support .htaccess (Apache) and web.config (IIS) files."
PLG_SYSTEM_HTTPHEADER_MESSAGE_STATICHEADERS_NOT_WRITTEN_NO_SERVER_CONFIGFILE_FOUND="We couldn't find any active .htaccess or web.config file to apply the rules to. Please first rename the htaccess.txt to .htaccess or web.config.txt to web.config and make sure the configuration files works."
PLG_SYSTEM_HTTPHEADER_MESSAGE_STATICHEADERS_WRITTEN="All static headers have been written to the server configuration file."
PLG_SYSTEM_HTTPHEADER_MESSAGE_STATICHEADERS_NOT_WRITTEN="Your %1$s file is not writable or there was a problem creating the file. You will have to upload the following code by hand. Select the following code and then paste into a new text file. Name this file '%1$s' and upload it to your site root folder.<br><pre>%2$s</pre>"
```

zero-24 committed:
improve en-GB thanks @brianteeman

zero-24 commented:
Thanks @brianteeman I have just updated the branch and the PR description. Thanks.

ReLater commented on httpheader.php:
```diff
@@ -2,12 +2,18 @@
 defined('_JEXEC') or die;
 
+use Joomla\CMS\Application\CMSApplication;
+use Joomla\CMS\Language\Text;#
```
Remove trailing `#`

zero-24 replied:
Thanks just did that change in my development branch. It will be fixed by the next commit in the comming days as I'm about to allocate time to adress a bruch of things that @yvesh already reported to me in the next week. Thanks for your report @ReLater

zero-24 replied:
resolved now

pe7er commented:
@zero-24 I've added the new nl-NL strings with PR: https://github.com/zero-24/plg_system_httpheader/pull/8

zero-24 commented:
Thanks I will only apply alpha ordering of the strings when i get back on the code, hopefully tomorrow :-)

zero-24 committed:
merge nl-NL, fist round of fixes suggested by @yvesh

zero-24 committed:
seccond round of @yvesh fixes

zero-24 committed:
set php minimum back to 7.0

zero-24 committed:
add server software checks

zero-24 committed:
allow the plugin to be installed in 3.8 too

zero-24 committed:
array() vs [] and statucHeaderConfiguration global var and additional use calls

SniperSister commented on httpheader.xml:
```diff
@@ -1,11 +1,11 @@
 <extension version="3.9" type="plugin" group="system" method="upgrade">
 	<name>plg_system_httpheader</name>
-	<creationDate>09.12.2018</creationDate>
+	<creationDate>16.02.2019</creationDate>
```
I would suggest using an international date format

zero-24 replied:
fixed

brianteeman replied:
```suggestion
	<creationDate>Feb 2019</creationDate>
```

brianteeman replied:
the majority of creationDate are in this format

zero-24 replied:
fixed thanks

SniperSister commented on httpheader.php:
```diff
@@ -47,7 +53,39 @@ class PlgSystemHttpHeader extends JPlugin
+	 */
+	const SERVER_CONFIG_FILE_HTACCESS = '.htaccess';
+
+		/**
```
Indent

zero-24 replied:
fixed

SniperSister commented on httpheader.php:
```diff
@@ -68,106 +106,493 @@ public function onAfterInitialise()
+		// Write the static headers
+		$result = $this->writeStaticHeaders();
+
+		if ($result)
```
Personally I would handle it the other way around: handle errors instead of success case

zero-24 replied:
I handle both cases. Just a few lines later the error message is thrown. Or do you want me to handle only the error cases?

SniperSister replied:
I personally handle error cases in the if and let the success case bubble through the function all the way down to the final return - but that's personal style and you can happily ignore it :)

zero-24 replied:
Ah got it now, makes sense. It is on my list to be changed. Thanks!

zero-24 replied:
Fixed by: https://github.com/zero-24/plg_system_httpheader/pull/6/commits/b35656c37d6a96e371610ba541f379a0a302b777 thanks.

SniperSister commented on httpheader.php:
```diff
@@ -68,106 +106,493 @@ public function onAfterInitialise()
+	 */
+	private function getServerConfigFilePath($file)
+	{
+		if (self::SERVER_CONFIG_FILE_HTACCESS === $file)
```
a simple return JPATH_ROOT . '/' . $file should do the job too, you aren't catching the "false" case anyway

zero-24 replied:
fixed

SniperSister commented on httpheader.php:
```diff
@@ -68,106 +106,493 @@ public function onAfterInitialise()
+	 *
+	 * @since   1.0.6
+	 */
+	private function getStaticHeaderConfiguration($pluginParams = false)
```
It's a non static function, so params are available in $this->params anyway

zero-24 replied:
depending on which state this method is used i use different sets of the params. One time (the normal case) I use $this->params for the time where we set them in the frontend.

I only get them manually in the case of onExtensionAfterSave as than this->params is not updated yet as far as i know. I might be wrong.

SniperSister replied:
Ah ok, makes sense!

SniperSister commented on httpheader.php:
```diff
@@ -68,106 +106,493 @@ public function onAfterInitialise()
+		if ($pluginParams->get('xframeoptions'))
 		{
-			$this->app->setHeader('X-XSS-Protection', '1; mode=block');
+			$staticHeaderConfiguration['X-Frame-Options' . '#both'] = 'SAMEORIGIN';
```
I don't really get what "both" is doing here. It's related to the client, yes, but why hardcoding it?

zero-24 replied:
The point is when enabled this header should be applied to backend and frontend. There is no configuration option to only apply this on one site. For this reason this is hardcoded to be `both` ;)

zero-24 committed:
date format

zero-24 committed:
adress issues raised by @snipersister

zero-24 committed:
thanks @brianteeman

brianteeman commented on httpheader.xml:
```diff
@@ -103,7 +111,7 @@
 					default="31536000"
 					filter="integer"
-					validate="options"
+					validate="number"
```
do you need the validate when you already have the filter? dont they both do the same thing?

zero-24 replied:
validate and filters do different things ;)

brianteeman commented on language/en-GB/en-GB.plg_system_httpheader.ini:
```diff
@@ -0,0 +1,50 @@
+COM_PLUGINS_CSP_FIELDSET_LABEL="Content-Security-Policy (CSP)"
+PLG_SYSTEM_HTTPHEADER="System - HTTPHeaders"
+PLG_SYSTEM_HTTPHEADER_ADDITIONAL_HEADER="Define additional HTTP Header"
+PLG_SYSTEM_HTTPHEADER_ADDITIONAL_HEADER_DESC="Using this you can set also different values from the default ones and also additional headers when it is supported. The supported headers are: <br><ul><li>Strict-Transport-Security</li><li>Content-Security-Policy</li><li>Content-Security-Policy-Report-Only</li><li>X-Frame-Options</li><li>X-XSS-Protection</li><li>X-Content-Type-Options</li><li>Referre ...
```
```suggestion
PLG_SYSTEM_HTTPHEADER_ADDITIONAL_HEADER_DESC="Using this you can set different values from the default ones and also additional headers when it is supported. The supported headers are: <br><ul><li>Strict-Transport-Security</li><li>Content-Security-Policy</li><li>Content-Security-Policy-Report-Only</li><li>X-Frame-Options</li><li>X-XSS-Protection</li><li>X-Content-Type-Options</li><li>Referrer-Policy</li><li>Expect-CT</li><li>Feature-Policy</li></ul>"
```

brianteeman commented on language/en-GB/en-GB.plg_system_httpheader.ini:
```diff
@@ -0,0 +1,50 @@
+PLG_SYSTEM_HTTPHEADER_ADDITIONAL_HEADER_DESC="Using this you can set also different values from the default ones and also additional headers when it is supported. The supported headers are: <br><ul><li>Strict-Transport-Security</li><li>Content-Security-Policy</li><li>Content-Security-Policy-Report-Only</li><li>X-Frame-Options</li><li>X-XSS-Protection</li><li>X-Content-Type-Options</li><li>Referre ...
+PLG_SYSTEM_HTTPHEADER_ADDITIONAL_HEADER_KEY="HTTP Header"
+PLG_SYSTEM_HTTPHEADER_ADDITIONAL_HEADER_VALUE="HTTP Header Value"
+PLG_SYSTEM_HTTPHEADER_CONTENTSECURITYPOLICY="<a href='https://scotthelme.co.uk/content-security-policy-an-introduction' target='_blank'>Content Security Policy (CSP)</a>"
```
as this is a target blank then it should have a rel=noopener noreferrer
yes I know scott's site will be safe but others may see this and copy it

zero-24 replied:
done that for all links and all languages now thanks.

brianteeman commented on language/en-GB/en-GB.plg_system_httpheader.ini:
```diff
@@ -0,0 +1,50 @@
+PLG_SYSTEM_HTTPHEADER_ADDITIONAL_HEADER_KEY="HTTP Header"
+PLG_SYSTEM_HTTPHEADER_ADDITIONAL_HEADER_VALUE="HTTP Header Value"
+PLG_SYSTEM_HTTPHEADER_CONTENTSECURITYPOLICY="<a href='https://scotthelme.co.uk/content-security-policy-an-introduction' target='_blank'>Content Security Policy (CSP)</a>"
+PLG_SYSTEM_HTTPHEADER_CONTENTSECURITYPOLICY_DESC="With this option all values for the 'Content-Security-Policy' can be set individually"
```
```suggestion
PLG_SYSTEM_HTTPHEADER_CONTENTSECURITYPOLICY_DESC="With this option all values for the 'Content-Security-Policy' can be set individually."
```

brianteeman commented on language/en-GB/en-GB.plg_system_httpheader.ini:
```diff
@@ -0,0 +1,50 @@
+PLG_SYSTEM_HTTPHEADER_ADDITIONAL_HEADER_VALUE="HTTP Header Value"
+PLG_SYSTEM_HTTPHEADER_CONTENTSECURITYPOLICY="<a href='https://scotthelme.co.uk/content-security-policy-an-introduction' target='_blank'>Content Security Policy (CSP)</a>"
+PLG_SYSTEM_HTTPHEADER_CONTENTSECURITYPOLICY_DESC="With this option all values for the 'Content-Security-Policy' can be set individually"
+PLG_SYSTEM_HTTPHEADER_CONTENTSECURITYPOLICY_REPORT_ONLY_DESC="This option uses the header 'Content-Security-Policy-Report-Only' instead of 'Content-Security-Policy"
```
```suggestion
PLG_SYSTEM_HTTPHEADER_CONTENTSECURITYPOLICY_REPORT_ONLY_DESC="This option uses the header 'Content-Security-Policy-Report-Only' instead of 'Content-Security-Policy'."
```

brianteeman commented on language/en-GB/en-GB.plg_system_httpheader.ini:
```diff
@@ -0,0 +1,50 @@
+PLG_SYSTEM_HTTPHEADER_CONTENTSECURITYPOLICY_VALUES_VALUE="Value"
+PLG_SYSTEM_HTTPHEADER_HEADER_CLIENT="Client"
+PLG_SYSTEM_HTTPHEADER_HEADER_CLIENT_BOTH="Both"
+PLG_SYSTEM_HTTPHEADER_HSTS="<a href='https://hstspreload.org' target='_blank'>HTTP Strict Transport Security (HSTS)</a>"
```
as before with the noopneer noreferrer

brianteeman commented on language/en-GB/en-GB.plg_system_httpheader.ini:
```diff
@@ -0,0 +1,50 @@
+PLG_SYSTEM_HTTPHEADER_HEADER_CLIENT="Client"
+PLG_SYSTEM_HTTPHEADER_HEADER_CLIENT_BOTH="Both"
+PLG_SYSTEM_HTTPHEADER_HSTS="<a href='https://hstspreload.org' target='_blank'>HTTP Strict Transport Security (HSTS)</a>"
+PLG_SYSTEM_HTTPHEADER_HSTS_DESC="This option activates 'Strict Transport Security' and allows the configuration of the values"
```
```suggestion
PLG_SYSTEM_HTTPHEADER_HSTS_DESC="This option activates 'Strict Transport Security' and allows the configuration of the values."
```

brianteeman commented on language/en-GB/en-GB.plg_system_httpheader.ini:
```diff
@@ -0,0 +1,50 @@
+PLG_SYSTEM_HTTPHEADER_HSTS="<a href='https://hstspreload.org' target='_blank'>HTTP Strict Transport Security (HSTS)</a>"
+PLG_SYSTEM_HTTPHEADER_HSTS_DESC="This option activates 'Strict Transport Security' and allows the configuration of the values"
+PLG_SYSTEM_HTTPHEADER_HSTS_MAXAGE="Maximum registration time"
+PLG_SYSTEM_HTTPHEADER_HSTS_MAXAGE_DESC="This option sets the time for 'max-age', it is specified in seconds. The default value is 31536000, which corresponds to one year"
```
```suggestion
PLG_SYSTEM_HTTPHEADER_HSTS_MAXAGE_DESC="This option sets the time in seconds for 'max-age'. The default value is 31536000, which corresponds to one year."
```

brianteeman commented on language/en-GB/en-GB.plg_system_httpheader.ini:
```diff
@@ -0,0 +1,50 @@
+PLG_SYSTEM_HTTPHEADER_HSTS_MAXAGE="Maximum registration time"
+PLG_SYSTEM_HTTPHEADER_HSTS_MAXAGE_DESC="This option sets the time for 'max-age', it is specified in seconds. The default value is 31536000, which corresponds to one year"
+PLG_SYSTEM_HTTPHEADER_HSTS_PRELOAD="Preload"
+PLG_SYSTEM_HTTPHEADER_HSTS_PRELOAD_DESC="This option activates the opt-in for inclusion in so-called browser preload lists"
```
```suggestion
PLG_SYSTEM_HTTPHEADER_HSTS_PRELOAD_DESC="This option activates the opt-in for inclusion in so-called browser preload lists."
```

brianteeman commented on language/en-GB/en-GB.plg_system_httpheader.ini:
```diff
@@ -0,0 +1,50 @@
+PLG_SYSTEM_HTTPHEADER_WRITE_STATIC_HEADERS_DESC="When enabled the headers (excluding the Content-Security-Policy) will be added to the existing server configuration file as soon as you save this extension. This way the headers will be enforced from the server side. At this time we only support .htaccess (Apache) and web.config (IIS) files."
+PLG_SYSTEM_HTTPHEADER_XCONTENTTYPEOPTIONS="<a href='https://scotthelme.co.uk/hardening-your-http-response-headers/#x-content-type-options' target='_blank'>X-Content-Type-Options</a>"
+PLG_SYSTEM_HTTPHEADER_XFRAMEOPTIONS="<a href='https://scotthelme.co.uk/hardening-your-http-response-headers/#x-frame-options' target='_blank'>X-Frame-Options</a>"
+PLG_SYSTEM_HTTPHEADER_XML_DESCRIPTION="This Plugin helps you to set the HTTP Security Headers"
```
```suggestion
PLG_SYSTEM_HTTPHEADER_XML_DESCRIPTION="This Plugin helps you to set the HTTP Security Headers."
```

brianteeman commented on language/en-GB/en-GB.plg_system_httpheader.sys.ini:
```diff
@@ -1,45 +1,5 @@
-PLG_SYSTEM_HTTPHEADER_XML_DESCRIPTION="This Plugin helps you to set the HTTP Security Headers"
-PLG_SYSTEM_HTTPHEADER_XXSSPROTECTION="<a href='https://scotthelme.co.uk/hardening-your-http-response-headers/#x-xss-protection' target='_blank'>X-XSS-Protection</a>"
\ No newline at end of file
+PLG_SYSTEM_HTTPHEADER_XML_DESCRIPTION="This Plugin helps you to set the HTTP Security Headers"
```
```suggestion
PLG_SYSTEM_HTTPHEADER_XML_DESCRIPTION="This Plugin helps you to set the HTTP Security Headers."
```

zero-24 committed:
add suggestions by @brianteeman thanks :+1:

zero-24 commented:
Thanks @brianteeman for the language suggestions. I have just pushed them as a single commit your suggestions are very much appreciated. 👍

zero-24 committed:
catch the error and the success message is the default case

zero-24 committed:
the htacces gets lower priority else we have the headers two times

zero-24 committed:
phpcs

Sandra97 committed:
fr-FR strings by @Sandra97 

* Typos corrected - Wikipedia links edited

* New strings added

* Update fr-FR.plg_system_httpheader.ini

* Update fr-FR.plg_system_httpheader.sys.ini

* Typo corrected

Sandra97 committed:
Little improvements (#12)

Sandra97 committed:
Bring consistency with other plugin translations (#11)

zero-24 committed:
undefined variable

zero-24 committed:
correct getServerConfigFilePath

zero-24 committed:
correct getServerConfigFilePath

zero-24 committed:
correct getServerConfigFile

zero-24 committed:
push server config file to the success message

zero-24 committed:
update fr-FR to use noopener and noreferrer

zero-24 committed:
add extra space for the csp headers after the ;

jeckodevelopment committed:
Italian language file (#13)

* Italian language file

Italian language for the plugin.

* Update it-IT.plg_system_httpheader.sys.ini

Updated language file to match the new branch

* Rename it-IT.plg_system_httpheader.sys.ini to it-IT.plg_system_httpheader.ini

* Create it-IT.plg_system_httpheader.sys.ini

zero-24 committed:
noopener noreferrer

zero-24 committed:
use the same style for the translator mention

zero-24 committed:
pull of the updater file changes

zero-24 commented:
Ok merging this now. Thanks all for your help and feedback on this one 
@brianteeman
@pe7er
@SniperSister
@ReLater
@Sandra97
@jeckodevelopment

As soon as the com_csp PR (https://github.com/joomla/joomla-cms/pull/24365) is merged I can start preparing to move this to 4.0 too as the config is moved there and I don't want to run into conflicts where i miss to change an config string ;)

zero-24 commented:
Just released 1.0.6: https://github.com/zero-24/plg_system_httpheader/releases/tag/1.0.6 🎉
